### DISCUSSION TITLE
How to get cargo miri as a check

### DISCUSSION THREAD

**Discussion by TheNeikos** [ipetkov/crane#44] [Q&A] [answered]
I've looked at the cargoFmt command, but I am not sure how to get it to run. 

Did someone get it to work yet? I understand its a bit more 'difficult' because of the extra toolchain because miri is nightly only?

**Comment by ipetkov** [comment]
I haven't used `cargo miri` as of yet, so pardon the light speculation here, but I think as long as the toolchain contains the `miri` component and as long as `cargo miri` can safely run from inside of the Nix sandbox it should be as simple as:

1. Add the toolchain derivation as a `nativeBuildInput` for the crate's derivation
2. Call `cargo miri` in a test phase

I did notice that the miri README mentioned:

```
The first time you run Miri, it will perform some extra setup and install some dependencies. It will ask you for confirmation before installing anything.
```

Not sure if this means miri needs to install its own additional components from the network the first time it runs. It would be interesting if they could be packaged up for general Nix consumption if that's the case...

**Reply by ipetkov** [reply]
Okay, judging from the output of `cargo +nightly miri setup` we need:

* A recent build of `xargo`. This is something that can be packaged since it's a cargo project
* The stdlib source, which needs to come from the `rust-src` component of the toolchain

**Comment by TheNeikos** [comment]
Alright! I got it working with the following steps:

- Acquire an unstable rust toolchain (I am using oxalicas rust overlay):
```rust
          unstableRustTarget = pkgs.rust-bin.selectLatestNightlyWith (toolchain: toolchain.default.override {
            extensions = [ "rust-src" "miri" ];
          });
```
- build the latest xargo (I am not show how miri checks what the 'latest' is, but it is needed or it will try downloading?)
```
         xargo = pkgs.rustPlatform.buildRustPackage
            rec {
              pname = "xargo";
              version = "0.3.26";

              src = pkgs.fetchFromGitHub {
                owner = "japaric";
                repo = pname;
                rev = "v${version}";
                hash = "sha256-MPopR58EIPiLg79wxf3nDy6SitdsmuUCjOLut8+fFJ4=";
              };

              cargoHash = "sha256-LmOu7Ni6TkACHy/ZG8ASG/P2UWEs3Qljz4RGSW1i3zk=";

              doCheck = false;

              buildInputs = [ ];
              nativeBuildInputs = [ ];
            };
```
- set the `XARGO_RUST_SRC` env variable to `${unstableRustTarget}/lib/rustlib/src/rust/library`
- Run `cargo miri test` !
- Enjoy!

**Reply by TheNeikos** [reply]
Now, the last step is to get `lib.cargoBuild` to work together with it.

The only thing I think can be problematic, is that the xargo version and the miri version need to match up somehow. I am not sure how this can be automated...

**Reply by ipetkov** [reply]
I'd be open to packaging up `xargo` locally within `crane`, I'd anticipate that as long as we have a recent `xargo` version miri should just work

**Reply by ipetkov** [reply]
I tried out adding a wrapper which pulls in `xargo`, but the problem I ran into is that `xargo` does not play nicely with Nix's sandbox. it appears to insist on [grabbing a lock in the sysroot](https://github.com/japaric/xargo/blob/6604e48fd48876211fd4a32dd6ed2213e9bb0147/src/xargo.rs#L68) which crashes on any invocation

**Reply by JeremiahSecrist** [reply]
maybe a temp directory could be used in which the sysroot data is copied over to satisfy this constraint? I recently got interested in miri for extra testing for my projects and found this thread. I hope raising a 2 year old conversation isn't an issue.

**Comment by tjjfvi** [comment, marked as answer]
For anyone looking to do this now that miri no longer uses `xargo`, I was able to get it working with the following:
```nix
checks.miri = craneLib.mkCargoDerivation (
  craneConfig # base config, with `src`, `cargoArtifacts`, etc.
  // {
    name = "miri-tests";
    buildPhaseCargoCommand = ''
      # test with miri
      cargo miri run ...
    '';
    doInstallCargoArtifacts = false;
    MIRI_SYSROOT = craneLib.mkCargoDerivation {
      pname = "miri-sysroot";
      version = "0.0.0";
      dontUnpack = true;
      buildPhaseCargoCommand = ''
        MIRI_SYSROOT=$out cargo miri setup
      '';
      dontFixup = true;
      cargoLock =
        pkgs.runCommand "sysroot-cargoLock"
          { nativeBuildInputs = [ rustToolchain ]; }
          ''
            cp `rustc --print sysroot`/lib/rustlib/src/rust/library/Cargo.lock $out
          '';
      cargoArtifacts = null;
    };
  }
);
```
